### PR TITLE
Add user profiles api

### LINKED ISSUE #48: Add API fror User profiles

### PR DESCRIPTION
**Pull request by BartoszChrostowski**
image

image

Closes #48

### CODE REVIEW

File: backend/TreeOfAKind.API/TreeOfAKind.API.csproj [hunk 1]
```diff
@@ -36,6 +36,7 @@
     <PackageReference Include="MediatR" Version="9.0.0" />
     <PackageReference Include="Microsoft.Extensions.Configuration.UserSecrets" Version="3.1.9" />
     <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
+    <PackageReference Include="NodaTime.Serialization.SystemTextJson" Version="1.0.0" />
```
Comment (Dragemil) [inline backend/TreeOfAKind.API/TreeOfAKind.API.csproj:39]: Another serialization package?

File: backend/TreeOfAKind.API/UserProfiles/UserProfileController.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        /// <response code="201">Returns uuid of created user profile</response>
+        /// <response code="400">Command is not valid</response>
+        /// <response code="401">User is not authenticated</response>
+        [HttpPost]
+        [Authorize]
+        [ProducesResponseType(typeof(IdDto), StatusCodes.Status201Created)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status409Conflict)]
+        [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
```
Comment (Dragemil) [inline backend/TreeOfAKind.API/UserProfiles/UserProfileController.cs:52]: That isn't in docs.

File: backend/TreeOfAKind.API/UserProfiles/UserProfileController.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+        
+        
+        /// <summary>
+        /// Gets a User Profile of an authenticated user (from token).
+        /// </summary>
+        /// <returns>User profile of an authenticated user</returns>
+        /// <response code="200">Returns user profile</response>
+        /// <response code="401">User is not authenticated</response>
+        /// <response code="404">User profile has not been created</response>
```
Comment (Dragemil) [inline backend/TreeOfAKind.API/UserProfiles/UserProfileController.cs:75]: Queries never fail, they can only return something, and that something might be `null`.

File: backend/TreeOfAKind.API/UserProfiles/UserProfileController.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        ///     {
+        ///        "firstName": "Bartosz",
+        ///        "lastName": "Chrostowski",
+        ///        "birthDate": "1998-02-12T00:00:00.000Z"
+        ///     }
+        ///
+        /// </remarks>
+        /// <param name="request"></param>
+        /// <returns>Uuid of created user profile</returns>
+        /// <response code="201">Returns uuid of created user profile</response>
```
Comment (Dragemil) [inline backend/TreeOfAKind.API/UserProfiles/UserProfileController.cs:43]: That's not very command-like.

Comment (BartoszChrostowski) [reply backend/TreeOfAKind.API/UserProfiles/UserProfileController.cs:43]: image
[ling](https://vladikk.com/2017/03/20/tackling-complexity-in-cqrs/)

Comment (BartoszChrostowski) [reply backend/TreeOfAKind.API/UserProfiles/UserProfileController.cs:43]: Also, don't treat api as strictly cqrs compliant

I suppose there may be some non-cqrs endpoints in the future and nothing's bad with it

File: backend/TreeOfAKind.API/UserProfiles/UserProfileController.cs (hunk 3 above)
Comment (BartoszChrostowski) [reply backend/TreeOfAKind.API/UserProfiles/UserProfileController.cs:75]: idk man, thats another enforcing full cqrs on api, but ok to me there's no difference between 404 (not found) and null but whatever, we can return null by convention

File: backend/TreeOfAKind.API/SeedWork/BusinessRuleValidationExceptionProblemDetails.cs [hunk 5]
```diff
@@ -1,16 +1,25 @@
...
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
 using TreeOfAKind.Domain.SeedWork;
+using TreeOfAKind.Domain.UserProfiles;
 
 namespace TreeOfAKind.API.SeedWork
 {
     public class BusinessRuleValidationExceptionProblemDetails : Microsoft.AspNetCore.Mvc.ProblemDetails
     {
+        private static readonly Dictionary<Type, int> BusinessRuleToStatus = new Dictionary<Type, int>()
```
[the remark this answers is not in the document]
Comment (BartoszChrostowski) [reply backend/TreeOfAKind.API/SeedWork/BusinessRuleValidationExceptionProblemDetails.cs:11]: why not 409?

[the remark this answers is not in the document]
Comment (BartoszChrostowski) [reply backend/TreeOfAKind.API/SeedWork/BusinessRuleValidationExceptionProblemDetails.cs:11]: will add 422 on every

File: backend/TreeOfAKind.API/UserProfiles/UserProfileController.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            var authUserId = User.FindFirst(ClaimTypes.NameIdentifier).Value;
+
+            var userId = await _mediator.Send(
+                new CreateOrUpdateUserProfileCommand(
+                    authUserId,
+                    request.FirstName,
+                    request.LastName,
+                    request.BirthDate));
+
+            return Created(string.Empty, new IdDto {Id = userId.Value});
```
Comment (Dragemil) [inline backend/TreeOfAKind.API/UserProfiles/UserProfileController.cs:64]: Please provide a generic command result DTO that has a success `bool`, an entityId as `string?`, and maybe later some validation errors.

**Review by Dragemil — CHANGES_REQUESTED** [review]

**Pull request merged**